File: jobs/update_users.py
import common
import time

def update_users():
    start = time.time()
    
    users_pg = common.pb_get_records('users', {'perPage': 10})
    while users_pg['page'] <= users_pg['totalPages']:
        process_users(users_pg['items'])
        users_pg = common.pb_get_records('users', {'perPage': 10, 'page': users_pg['page'] + 1})
        
    print('update_users finished in', round(time.time() - start, 2), 'seconds')
    
    
def process_users(users:list):
    for user in users:
        user_id = user['id']
        update_stats(user_id)

def update_stats(user_id:str):
    
    past_picks = common.pb_get_records('picks', {'perPage': 1000, 'filter': f'user="{user_id}" && status="past"'})
    win_picks = list(filter(lambda p: p['result'] == 'W', past_picks['items']))
    lose_picks = list(filter(lambda p: p['result'] == 'L', past_picks['items']))
    total_picks = past_picks['totalItems']
    win_loss_ratio = len(win_picks) / len(lose_picks) if lose_picks else len(win_picks)
    win_pick_rate = 0
    if total_picks:
        win_pick_rate = int(len(win_picks) / total_picks * 100)
    
    latest_stats = {
        'total_picks': total_picks,
        'win_picks': len(win_picks),
        'lose_picks': len(lose_picks),
        'win_loss_ratio': win_loss_ratio,
        'win_pick_rate': win_pick_rate
    }
    
    stats_record = common.pb_get_first_item('stats', f'user="{user_id}"')
    if not stats_record:
        new_stats = common.pb_create_record('stats', {
            'user': user_id,
            **latest_stats
        })
        common.pb_update_record('users', user_id, {'stats': new_stats['id']})
        print('Created stats for user', user_id)
        return
    
    common.pb_update_record('stats', stats_record['id'], latest_stats)
    print('Updated stats for user', user_id)
    
            
def get_job():
    return {
        'name': 'update_users',
        'description': 'Updates all users to reflect their current stats.',
        'function': update_users
    }

    
if __name__ == '__main__':
    update_users()

File: src/pages/api/cron/update_users.ts
import type { APIRoute } from 'astro';
import { getAPB } from '@/lib/data';
import { CRON_SECRET } from 'astro:env/server';
import { getStatsByUserId } from '@/lib/stats';
import { getPicksByUser } from '@/lib/picks';

interface UserStats {
  total_picks: number;
  win_picks: number;
  lose_picks: number;
  win_loss_ratio: number;
  win_pick_rate: number;
}

async function attachStatsToUser(userId: string, statsId) {
  const pb = getAPB();

  await pb.collection('users').update(userId, {
    stats: statsId,
  });
}

async function updateStats(userId: string) {
  const pb = getAPB();
  try {
    const pastPicks = await getPicksByUser(userId, 'past');

    const winPicks = pastPicks.filter((p: any) => p.result === 'W');
    const losePicks = pastPicks.filter((p: any) => p.result === 'L');
    const totalPicks = pastPicks.length;

    const winLossRatio = losePicks.length
      ? winPicks.length / losePicks.length
      : winPicks.length;
    const winPickRate = totalPicks
      ? Math.round((winPicks.length / totalPicks) * 100)
      : 0;

    const latestStats: UserStats = {
      total_picks: totalPicks,
      win_picks: winPicks.length,
      lose_picks: losePicks.length,
      win_loss_ratio: winLossRatio,
      win_pick_rate: winPickRate,
    };

    // Get or create stats record
    const statsRecord = await getStatsByUserId(userId);
    if (statsRecord) {
      await pb.collection('stats').update(statsRecord.id, {
        user: userId,
        ...latestStats,
      });
      return { action: 'UPDATED', userId };
    }

    const newStatsRecord = await pb.collection('stats').create(latestStats);
    await attachStatsToUser(userId, newStatsRecord.id);

    return { action: 'CREATED', userId };
  } catch (error) {
    console.error('Error updating stats for user', userId, error);
    return { action: 'FAILED', userId, error: error.message };
  }
}

export const POST: APIRoute = async ({ request }) => {
  if (request.headers.get('Cron-Secret') !== CRON_SECRET) {
    return new Response('Unauthorized', { status: 401 });
  }

  try {
    const startTime = Date.now();
    const pb = getAPB();

    const users = await pb.collection('users').getFullList();
    const results = [];

    for (const user of users) {
      const res = await updateStats(user.id);
      results.push(res);
    }

    const createdCount = results.filter((r) => r.action === 'CREATED').length;
    const updatedCount = results.filter((r) => r.action === 'UPDATED').length;
    const failed = results.filter((r) => r.action === 'FAILED');
    const failedCount = failed.length;

    const duration = (Date.now() - startTime) / 1000;

    return new Response(
      JSON.stringify(
        {
          message: 'Update users job completed',
          duration: `${duration.toFixed(2)} seconds`,
          stats: {
            created: createdCount,
            updated: updatedCount,
            failed: failedCount,
          },
          results,
        },
        null,
        2
      ),
      {
        status: 200,
      }
    );
  } catch (error) {
    console.error('Error updating users:', error);
    return new Response(JSON.stringify({ error: 'Failed to update users' }), {
      status: 500,
    });
  }
};
